### ISSUE TITLE
model.trainable=False does nothing in tensorflow keras

### ISSUE THREAD

**Issue by Nitinsiwach** [tensorflow/tensorflow#29535]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):source
- TensorFlow version (use command below):1.13
- Python version:2.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:10.1.105
- GPU model and memory:m60,8gb

It seems setting `model.trainable=False` in tensorflow keras does nothing except for to print a wrong model.summary(). Here is the code to reproduce the issue:

```
import tensorflow as tf
import numpy as np
IMG_SHAPE = (160, 160, 3)

# Create the base model from the pre-trained model MobileNet V2
base_model = tf.keras.applications.MobileNetV2(input_shape=IMG_SHAPE,
                                               include_top=False, 
                                               weights='imagenet')
base_model.trainable = False
# for layer in base_model.layers:
#     layer.trainable=False
bc=[] #before compile
ac=[] #after compile
for layer in base_model.layers:
    bc.append(layer.trainable)
print(np.all(bc)) #True
print(base_model.summary()) ##this changes to show no trainable parameters but that  is wrong given the output to previous np.all(bc)
base_model.compile(optimizer=tf.keras.optimizers.Adam(lr=0.001), 
              loss='categorical_crossentropy', 
              metrics=['accuracy'])
for layer in base_model.layers:
    ac.append(layer.trainable)
print(np.all(ac)) #True
print(base_model.summary()) #this changes to show no trainable parameters but that  is wrong given the output to previous np.all(ac)
```

**Comment by KabirKwatra** [comment]
tf.keras.Model documentation does not mention this attribute.

```python
>>> hasattr(base_model, 'trainable')
True
```
tf.keras.Model inherits attribute from Network.
Attribute seems to be defined here:
https://github.com/tensorflow/tensorflow/blob/9699b90e16b3538168dde330e1ac8745fe8b73b8/tensorflow/python/keras/engine/network.py#L194-L195

**Comment by KabirKwatra** [comment]
Setting the `trainable` attribute to False affects the models' `trainable_weights` attribute.

**Comment by KabirKwatra** [comment]
https://github.com/tensorflow/tensorflow/blob/c572ea2806176d036e4d81b5de22f7251449d371/tensorflow/python/keras/engine/base_layer.py#L766-L772

This Property seems to be the cause

**Comment by KabirKwatra** [comment]
When trainable is set to false, the trainable_weights property resolve to empty list. This causes the layer summary to display the parameter count to display incorrectly.

https://github.com/tensorflow/tensorflow/blob/9699b90e16b3538168dde330e1ac8745fe8b73b8/tensorflow/python/keras/utils/layer_utils.py#L234-L239

**Comment by Nitinsiwach** [comment]
setting `model.trainable=False` is the suggested way in even the Tensorflow documentation for transfer learning https://www.tensorflow.org/tutorials/images/transfer_learning

**Comment by Nitinsiwach** [comment]
> When trainable is set to false, the trainable_weights property resolve to empty list. This causes the layer summary to display the parameter count to display incorrectly.
> It is unclear to me how trainable_weights being set to an empty list makes layer summary to display the parameter count incorrectly. Can you explain a bit more please?
> https://github.com/tensorflow/tensorflow/blob/9699b90e16b3538168dde330e1ac8745fe8b73b8/tensorflow/python/keras/utils/layer_utils.py#L234-L239

**Comment by KabirKwatra** [comment]
I think the problem is actually that the individual layers in the model stay trainable when the model is not trainable.

```python
>>> import tensorflow as tf
>>> base_model = tf.keras.applications.MobileNetV2(input_shape=IMG_SHAPE,
                                               include_top=False, 
                                               weights='imagenet')
>>> base_model.trainable = False
>>> base_model.layers[0].trainable
True
```

**Comment by Nitinsiwach** [comment]
i mentioned that in the code bit i posted as minimal reproducible example. All layers remain trainable both before and after compiling the model

**Comment by achandraa** [comment]
Tried on Colab with TF 1.13.1 and was able to get mentioned model.summary with trainable params as 0.

**Comment by Nitinsiwach** [comment]
I can confirm that all three `model.trainable=False`, 'layer.trainable=False for layer in model.layers` and 'model.trainable=True` result in different behavior

**Comment by KabirKwatra** [comment]
https://github.com/tensorflow/tensorflow/blob/c572ea2806176d036e4d81b5de22f7251449d371/tensorflow/python/keras/engine/base_layer.py#L750-L754

This observer pattern does not seem to be functioning as expected.

**Comment by Nitinsiwach** [comment]
Any update on this?

**Comment by tanzhenyu** [comment]
This is fixed in 1.14, can you update the version and try it again?

**Comment by jvishnuvardhan** [comment]
@Nitinsiwach Please check a gist with 1.14 [here](https://colab.sandbox.google.com/gist/jvishnuvardhan/936cbc0184d3ae96bdc682f34a32d9a2/tf29535_model_tranables.ipynb). Thanks!

**Comment by jvishnuvardhan** [comment]
Automatically closing this out since I understand it to be resolved, but please let me know if I'm mistaken.Thanks!

**Comment by Aladoro** [comment]
Encountered this error in tensorflow 1.14 from a colab gpu runtime with the following code:

`import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import (Dense, Activation, BatchNormalization, 
                                     Reshape, Conv2D, LeakyReLU, Input)
from tensorflow.keras.optimizers import Adam

print(tf.__version__)

network = Sequential([
          Conv2D(64, 4, strides=(2, 2), padding='same', input_shape=(28,28,1)),
          LeakyReLU(),
          Conv2D(128, 4, strides=(2, 2), padding='same'),
          BatchNormalization(axis=-1),
          LeakyReLU(),
          Reshape((7*7*128,)),
          Dense(1024),
          BatchNormalization(axis=-1),
          LeakyReLU(),
          Dense(1),
          Activation('sigmoid')
      ])

network.compile(loss='binary_crossentropy', optimizer=Adam(2e-4, beta_1=0.5))
network.summary()
network.trainable = False
network.summary()`

**Comment by jvishnuvardhan** [comment]
@aladoro Please create a new issue by providing your platform details and the standalone code. Thanks!

**Comment by don-tpanic** [comment]
```
from keras.models import Model
from keras.applications.vgg16 import VGG16

model = VGG16(weights='imagenet', include_top=True, input_shape=(224, 224, 3))
model.trainable = False

model.summary()

model = Model(inputs=model.input, outputs=model.output)

model.summary()
```
This first summary shows there is no trainable weights. Wheras second summary shows all weights are trainable.

**Comment by jvishnuvardhan** [comment]
@don-tpanic Please create a new issue with more details on the issue and also provide platform details, version used etc. Thanks!

**Comment by babaish** [comment]
> ```
> from keras.models import Model
> from keras.applications.vgg16 import VGG16
> 
> model = VGG16(weights='imagenet', include_top=True, input_shape=(224, 224, 3))
> model.trainable = False
> 
> model.summary()
> 
> model = Model(inputs=model.input, outputs=model.output)
> 
> model.summary()
> ```
> 
> This first summary shows there is no trainable weights. Wheras second summary shows all weights are trainable.
> 
> ```
> for layer in model.layers:
>         layer.trainable = False
> ```
> 
> If doing it this way, both summaries show no trainable weights.

yes this is because confusingly, the behavior of model.trainable is different in keras vs tf.keras: in keras it only impact the variable of the model layer, without impacting the variable of all the sub layers, contrary to what happen in tf.keras. (Is it expected to have such different behavior between keras and tf.keras??)
In the second summary, you effectively creates a new model which set model.trainable = True by default, which explains the different behavior.
I find setting the trainable flag of the whole model very confusing, so I now only set it per sub layers. For example in tf.keras, if you do model.trainable = False and then model.layers[-4].trainable = True, the latter instruction has no effect as it does not change the trainable flag at the top of the model.

**Comment by Catadanna** [comment]
@babaish: Thank you, good point, useful to know!

**Comment by microprediction** [comment]
I also think the role of model.trainable could be clarified, and the overriding of trainability when the model is created. See this [gist](https://gist.github.com/microprediction/7aa395cec8a0ea2586936084a635cfba) showing that layer trainability is ignored. The behavior is a little counter-intuitive.
